src/osd/OSD.cc

```diff
@@ -7775,6 +7903,91 @@ void OSD::advance_pg(
       continue;
     }
 
+    unsigned new_pg_num =
+      (old_pg_num && nextmap->have_pg_pool(pg->pg_id.pool())) ?
+      nextmap->get_pg_num(pg->pg_id.pool()) : 0;
+    if (old_pg_num && new_pg_num && old_pg_num != new_pg_num) {
+      // check for merge
+      set<spg_t> children;
+      if (nextmap->have_pg_pool(pg->pg_id.pool())) {
+	spg_t parent;
+	if (pg->pg_id.is_merge(
+	      old_pg_num,
+	      new_pg_num,
+	      &parent)) {
+	  // we are merge source
+	  PGRef spg = pg; // carry a ref
+	  dout(20) << __func__ << " we are merge source, target is " << parent
+		   << dendl;
+	  pg->ch->flush();
```

Comment:
let's do a dout(1) here and before pg->merge_from(), since these may end up being expensive (esp. with filestore) and we'll want to be able to tell when they're happening if so